File: src/um/LoginToken.cc
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */

#include <sstream>

#include "LoginToken.h"
#include "NebulaUtil.h"
#include "ObjectXML.h"

using namespace std;

/* -------------------------------------------------------------------------- */
/* -------------------------------------------------------------------------- */
/* LoginTokenPool class                                                       */
/* -------------------------------------------------------------------------- */
/* -------------------------------------------------------------------------- */

const int LoginTokenPool::MAX_TOKENS = 10;

/* -------------------------------------------------------------------------- */

LoginTokenPool::~LoginTokenPool()
{
    reset();
}

/* -------------------------------------------------------------------------- */

void LoginTokenPool::reset()
{
    for (auto it = tokens.begin() ; it != tokens.end() ; ++it)
    {
        delete it->second;
    }

    tokens.clear();
}

/* -------------------------------------------------------------------------- */

int LoginTokenPool::set(std::string& utk, time_t valid, int egid)
{
    reset_expired(); // Expired token collector

    if (tokens.size() >= MAX_TOKENS || valid < -1 || valid == 0)
    {
        return -1;
    }

    LoginToken * tk = new LoginToken;

    utk = tk->set(utk, valid, egid);

    tokens.insert(std::pair<std::string, LoginToken *>(utk, tk));

    return 0;
}

/* -------------------------------------------------------------------------- */

int LoginTokenPool::reset(const std::string& utk)
{
    auto it = tokens.find(utk);

    if ( it == tokens.end() )
    {
        return -1;
    }

    delete it->second;

    tokens.erase(it);

    return 0;
}

/* -------------------------------------------------------------------------- */

void LoginTokenPool::reset_expired()
{
    for (auto it = tokens.begin(); it != tokens.end(); )
    {
        if ((it->second)->is_expired())
        {
            delete it->second;

            it = tokens.erase(it);
        }
        else
        {
            ++it;
        }
    }
}

/* -------------------------------------------------------------------------- */

bool LoginTokenPool::is_valid(const std::string& utk, int& egid, bool& exists_token)
{
    egid = -1;
    auto it = tokens.find(utk);

    if ( it == tokens.end() )
    {
        exists_token = false;
        return false;
    }

    exists_token = true;

    if ( it->second->is_valid(utk, egid) == true)
    {
        return true;
    }

    delete it->second;

    tokens.erase(it);

    return false;
}

/* -------------------------------------------------------------------------- */

void LoginTokenPool::from_xml_node(const std::vector<xmlNodePtr>& content)
{
    for (auto node : content)
    {
        LoginToken * tk = new LoginToken;
        std::string utk = tk->from_xml_node(node);

        tokens.insert(std::pair<std::string, LoginToken *>(utk, tk));
    }
}

/* -------------------------------------------------------------------------- */

std::string& LoginTokenPool::to_xml(std::string& xml) const
{
    std::ostringstream oss;

    for ( auto it = tokens.begin() ; it != tokens.end() ; ++it)
    {
        it->second->to_xml(oss);
    }

    xml = oss.str();

    return xml;
}

/* -------------------------------------------------------------------------- */
/* -------------------------------------------------------------------------- */
/* LoginToken class                                                           */
/* -------------------------------------------------------------------------- */
/* -------------------------------------------------------------------------- */

bool SessionToken::is_valid(const string& user_token) const
{
    return ((user_token == token) &&
            ((expiration_time == -1) || (time(0) < expiration_time)));
}

/* -------------------------------------------------------------------------- */

bool SessionToken::is_expired() const
{
    return ((time(0) > expiration_time) && (expiration_time != -1));
}

/* -------------------------------------------------------------------------- */

void SessionToken::reset()
{
    token.clear();
    expiration_time = 0;
}

/* -------------------------------------------------------------------------- */

const std::string& SessionToken::set(const std::string& user_token, time_t valid)
{
    if (valid == -1)
    {
        expiration_time = -1;
    }
    else if (valid > 0 )
    {
        expiration_time = time(0) + valid;
    }

    if (!user_token.empty())
    {
        token = user_token;
    }
    else
    {
        token = one_util::random_password();
    }

    return token;
}

/* -------------------------------------------------------------------------- */

void LoginToken::to_xml(std::ostringstream& xml) const
{
    xml << "<LOGIN_TOKEN>"
        << "<TOKEN>" << token << "</TOKEN>"
        << "<EXPIRATION_TIME>" << expiration_time << "</EXPIRATION_TIME>"
        << "<EGID>" << egid << "</EGID>"
        << "</LOGIN_TOKEN>";
}

/* -------------------------------------------------------------------------- */

const std::string& LoginToken::from_xml_node(const xmlNodePtr node)
{
    ObjectXML oxml(node);

    oxml.xpath(token, "/LOGIN_TOKEN/TOKEN", "");
    oxml.xpath<time_t>(expiration_time, "/LOGIN_TOKEN/EXPIRATION_TIME", 0);
    oxml.xpath<int>(egid, "/LOGIN_TOKEN/EGID", -1);

    return token;
}

/* -------------------------------------------------------------------------- */
/* -------------------------------------------------------------------------- */

File: src/fireedge/src/modules/containers/Settings/LoginToken/LoginToken.js
/* ------------------------------------------------------------------------- *
 * Copyright 2002-2025, OpenNebula Project, OpenNebula Systems               *
 *                                                                           *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may   *
 * not use this file except in compliance with the License. You may obtain   *
 * a copy of the License at                                                  *
 *                                                                           *
 * http://www.apache.org/licenses/LICENSE-2.0                                *
 *                                                                           *
 * Unless required by applicable law or agreed to in writing, software       *
 * distributed under the License is distributed on an "AS IS" BASIS,         *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
 * See the License for the specific language governing permissions and       *
 * limitations under the License.                                            *
 * ------------------------------------------------------------------------- */
import { SubmitButton, FormWithSchema, Tr, Translate } from '@ComponentsModule'
import { STYLE_BUTTONS, T } from '@ConstantsModule'
import { css } from '@emotion/css'
import {
  GroupAPI,
  useAuth,
  useAuthApi,
  useGeneralApi,
  UserAPI,
  useViews,
} from '@FeaturesModule'
import { useClipboard } from '@HooksModule'
import { timeToString } from '@ModelsModule'
import { FIELDS, SCHEMA } from '@modules/containers/Settings/LoginToken/schema'
import { useSettingWrapper } from '@modules/containers/Settings/Wrapper'
import { Box, Grid, Typography, useTheme } from '@mui/material'
import { Cancel, Check as CopiedIcon, Copy as CopyIcon } from 'iconoir-react'
import PropTypes from 'prop-types'
import { ReactElement, useCallback, useMemo } from 'react'
import { FormProvider, useForm } from 'react-hook-form'
import { v4 as uuidv4 } from 'uuid'

const styles = ({ typography, palette }) => ({
  buttonPlace: css({
    textAlign: 'center',
  }),
  buttonAction: css({
    width: '100%',
    marginBottom: typography.pxToRem(8),
  }),
  token: css({
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
  }),
  message: css({
    margin: `${typography.pxToRem(32)} 0 0`,
  }),
  tokenContainer: css({
    border: `1px solid ${palette.grey[300]}`,
    padding: typography.pxToRem(8),
    borderRadius: typography.pxToRem(8),
    marginBottom: typography.pxToRem(8),
  }),
})

const Row = ({ data = {}, groups = [], edit = () => undefined } = {}) => {
  const { copy, isCopied } = useClipboard()
  const theme = useTheme()
  const classes = useMemo(() => styles(theme), [theme])
  const { TOKEN = '', EXPIRATION_TIME = 0, EGID = '' } = data
  const groupToken =
    groups.find((group) => group?.ID === EGID)?.NAME || Tr(T.None)

  const handleCopy = useCallback(
    (evt) => {
      !isCopied && copy(TOKEN)
      evt.stopPropagation()
    },
    [copy, TOKEN, isCopied]
  )

  return (
    <>
      {TOKEN && EXPIRATION_TIME && EGID && (
        <Grid container role="row" className={classes.tokenContainer}>
          <Grid item sx={{ flexGrow: 1 }}>
            <Grid container spacing={2}>
              <Grid item xs={6}>
                <Typography variant="body2" gutterBottom sx={{ m: '1rem' }}>
                  <b>{`${Tr(T.ValidUntil)}: `}</b>
                  {timeToString(EXPIRATION_TIME)}
                </Typography>
              </Grid>
              <Grid item xs={6}>
                <Typography variant="body2" gutterBottom sx={{ m: '1rem' }}>
                  <b>{`${Tr(T.Group)}: `}</b>
                  {groupToken}
                </Typography>
              </Grid>
              <Grid item xs={12}>
                <Typography
                  variant="body2"
                  gutterBottom
                  sx={{ m: '1rem' }}
                  className={classes.token}
                >
                  <b>{`${Tr(T.Token)}: `}</b>
                  {TOKEN}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
          <Grid item>
            <SubmitButton
              icon={isCopied ? <CopiedIcon /> : <CopyIcon className="icon" />}
              onClick={handleCopy}
              aria-label="toggle password visibility"
            />
            <SubmitButton
              icon={<Cancel />}
              onClick={() => edit(TOKEN)}
              aria-label="delete"
            />
          </Grid>
        </Grid>
      )}
    </>
  )
}

Row.propTypes = {
  data: PropTypes.shape({
    TOKEN: PropTypes.string,
    EXPIRATION_TIME: PropTypes.string,
    EGID: PropTypes.string,
  }),
  groups: PropTypes.arrayOf(PropTypes.object),
  edit: PropTypes.func,
}

Row.displayName = 'LoginTokenRow'

/**
 * Section to change user configuration about UI.
 *
 * @returns {ReactElement} Settings configuration UI
 */
const LoginToken = () => {
  const { Legend, InternalWrapper } = useSettingWrapper()
  const { user } = useAuth()
  const { LOGIN_TOKEN = [], ID, NAME } = user
  const { data: groups = [], isLoading } = GroupAPI.useGetGroupsQuery()
  const [get, { data: lazyUserData }] = UserAPI.useLazyGetUserQuery()
  const { changeAuthUser } = useAuthApi()

  const userGroups = groups.filter((group) => {
    const arrayUsers = Array.isArray(group?.USERS?.ID)
      ? group?.USERS?.ID
      : [group?.USERS?.ID]

    return arrayUsers.includes(ID)
  })

  const arrayLoginToken = Array.isArray(LOGIN_TOKEN)
    ? LOGIN_TOKEN
    : [LOGIN_TOKEN]

  const { enqueueError } = useGeneralApi()
  const [updateUser] = UserAPI.useUpdateUserMutation()
  const [addLoginToken] = UserAPI.useLoginUserMutation()
  const { views, view: userView } = useViews()
  const theme = useTheme()
  const classes = useMemo(() => styles(theme), [theme])

  const { handleSubmit, ...methods } = useForm({
    reValidateMode: 'onChange',
    defaultValues: useMemo(
      () =>
        SCHEMA({ views, userView }).cast(
          {},
          {
            stripUnknown: true,
          }
        ),
      [LOGIN_TOKEN]
    ),
  })

  const handleDeleteLoginToken = useCallback(
    async (token) => {
      try {
        const newToken = await addLoginToken({ user: NAME, expire: 0, token })
        if (newToken?.data) {
          const newUserData = await get({ id: ID, __: uuidv4() })
          if (newUserData?.data) {
            changeAuthUser({
              ...user,
              LOGIN_TOKEN: newUserData.data?.LOGIN_TOKEN,
            })
          }
        }
      } catch {
        enqueueError(T.SomethingWrong)
      }
    },
    [updateUser, user, lazyUserData]
  )

  const handleCreateLoginToken = useCallback(
    async ({ EXPIRE: expire, EGID: gid } = {}) => {
      try {
        const newToken = await addLoginToken({ user: NAME, expire, gid })
        if (newToken?.data) {
          const newUserData = await get({ id: ID, __: uuidv4() })
          if (newUserData?.data) {
            changeAuthUser({
              ...user,
              LOGIN_TOKEN: newUserData.data?.LOGIN_TOKEN,
            })
          }
        }
      } catch {
        enqueueError(T.SomethingWrong)
      }
    },
    [addLoginToken, user, lazyUserData]
  )

  return (
    <Box>
      <Legend title={T.LoginToken} />
      <InternalWrapper>
        {!isLoading && (
          <>
            {arrayLoginToken.length > 0 && (
              <div>
                <div>
                  {arrayLoginToken.map((itemLoginToken, i) => (
                    <Row
                      data={itemLoginToken}
                      key={i}
                      groups={groups}
                      edit={handleDeleteLoginToken}
                    />
                  ))}
                </div>
              </div>
            )}
            <Box
              component="form"
              onSubmit={handleSubmit(handleCreateLoginToken)}
            >
              <FormProvider {...methods}>
                <FormWithSchema
                  cy={'logintoken-ui'}
                  fields={FIELDS(userGroups)}
                />
              </FormProvider>
              <Box className={classes.buttonPlace}>
                <SubmitButton
                  onClick={handleSubmit(handleCreateLoginToken)}
                  importance={STYLE_BUTTONS.IMPORTANCE.MAIN}
                  size={STYLE_BUTTONS.SIZE.MEDIUM}
                  type={STYLE_BUTTONS.TYPE.FILLED}
                  data-cy="addLoginToken"
                  label={T.GetNewToken}
                />
              </Box>
            </Box>
          </>
        )}
        <Typography
          variant="body2"
          gutterBottom
          sx={{ m: '1rem' }}
          className={classes.message}
        >
          <Translate word={T.MessageLoginToken} />
        </Typography>
      </InternalWrapper>
    </Box>
  )
}

export default LoginToken
